README.md
# Introduction

## AsyncStorage dependency is fixed, hurray!

This is a i18next cache layer to be used in the browser. It will load and cache resources from AsyncStorage and can be used in combination with the [chained backend](https://github.com/i18next/i18next-chained-backend).

# Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/i18next-async-storage-cache) or [downloaded](https://github.com/timbrandin/i18next-async-storage-cache/blob/master/i18nextAsyncStorageCache.min.js) from this repo.

```
# npm package
$ npm install i18next-async-storage-backend2
```

Wiring up with the chained backend:

```js
import i18next from 'i18next';
import Backend from 'i18next-chained-backend';
import AsyncStorageBackend from 'i18next-async-storage-backend2'; // primary use cache
import XHR from 'i18next-xhr-backend'; // fallback xhr load

i18next.use(Backend).init({
  backend: {
    backends: [
      AsyncStorageBackend, // primary
      XHR // fallback
    ],
    backendOptions: [
      {
        /* below options */
      },
      {
        loadPath: '/locales/{{lng}}/{{ns}}.json' // xhr load path for my own fallback
      }
    ]
  }
});
```

## Cache Backend Options

```js
{
  // prefix for stored languages
  prefix: 'i18next_res_',

  // expiration
  expirationTime: 7*24*60*60*1000,

  // language versions
  versions: {}
};
```

- Contrary to cookies behavior, the cache will respect updates to `expirationTime`. If you set 7 days and later update to 10 days, the cache will persist for 10 days

- Passing in a `versions` object (ex.: `versions: { en: 'v1.2', fr: 'v1.1' }`) will give you control over the cache based on translations version. This setting works along `expirationTime`, so a cached translation will still expire even though the version did not change. You can still set `expirationTime` far into the future to avoid this

Wiring up a service backend with the chained backend:

```js
import i18next from 'i18next';
import Backend from 'i18next-chained-backend';
import AsyncStorageBackend from 'i18next-async-storage-backend2'; // primary use cache
import ServiceBackend from 'i18next-service-backend'; // fallback service backend

const TRANSLATION_BACKEND = 'https://api.spacetranslate.com';
const TRANSLATION_BACKEND_PROJECTID = '[projectId]'; // i.e. [projectId].spacetranslate.com

i18next.use(Backend).init({
  backend: {
    backends: [
      AsyncStorageBackend, // primary
      ServiceBackend // fallback
    ],
    backendOptions: [
      {
        // prefix for stored languages
        prefix: 'i18next_res_',

        // expiration
        expirationTime: 7 * 24 * 60 * 60 * 1000,

        // language versions
        versions: {}
      },
      {
        service: TRANSLATION_BACKEND,
        projectId: TRANSLATION_BACKEND_PROJECTID,
        referenceLng: 'en',
        version: 'latest'
      }
    ]
  }
});
```


CHANGELOG.md
### 2.0.1

- Add type declarations

### 2.0.0

- Fix async-storage dependency

### 1.1.4

- fixes cache save call

### 1.1.1

- initial version


src/index.js
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as utils from './utils';

// get from whatever version of react native that is being used.

const storage = {
  setItem(key, value) {
    if (AsyncStorage) {
      return AsyncStorage.setItem(key, value);
    }
  },
  getItem(key, value) {
    if (AsyncStorage) {
      return AsyncStorage.getItem(key, value);
    }
    return undefined;
  }
};

function getDefaults() {
  return {
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

class Cache {
  constructor(services, options = {}) {
    this.init(services, options);

    this.type = 'backend';
  }

  init(services, options = {}) {
    this.services = services;
    this.options = utils.defaults(options, this.options || {}, getDefaults());
  }

  read(language, namespace, callback) {
    const nowMS = new Date().getTime();

    if (!AsyncStorage) {
      return callback(null, null);
    }

    storage
      .getItem(`${this.options.prefix}${language}-${namespace}`)
      .then(local => {
        if (local) {
          local = JSON.parse(local);
          if (
            // expiration field is mandatory, and should not be expired
            local.i18nStamp &&
            local.i18nStamp + this.options.expirationTime > nowMS &&
            // there should be no language version set, or if it is, it should match the one in translation
            this.options.versions[language] === local.i18nVersion
          ) {
            delete local.i18nVersion;
            delete local.i18nStamp;
            return callback(null, local);
          }
        }

        callback(null, null);
      })
      .catch(err => {
        console.warn(err);
        callback(null, null);
      });
  }

  save(language, namespace, data) {
    if (AsyncStorage) {
      data.i18nStamp = new Date().getTime();

      // language version (if set)
      if (this.options.versions[language]) {
        data.i18nVersion = this.options.versions[language];
      }

      // save
      storage.setItem(
        `${this.options.prefix}${language}-${namespace}`,
        JSON.stringify(data)
      );
    }
  }
}

Cache.type = 'backend';

export default Cache;


rollup.config.js
import babel from 'rollup-plugin-babel';
import uglify from 'rollup-plugin-uglify';
import nodeResolve from 'rollup-plugin-node-resolve';
import { argv } from 'yargs';

const format = argv.format || argv.f || 'iife';
const compress = argv.uglify;

const babelOptions = {
  exclude: 'node_modules/**',
  presets: [['es2015', { modules: false }], 'stage-0'],
  babelrc: false
};

const dest = {
  amd: `dist/amd/i18nextAsyncStorageBackend${compress ? '.min' : ''}.js`,
  umd: `dist/umd/i18nextAsyncStorageBackend${compress ? '.min' : ''}.js`,
  iife: `dist/iife/i18nextAsyncStorageBackend${compress ? '.min' : ''}.js`
}[format];

export default {
  entry: 'src/index.js',
  format,
  plugins: [babel(babelOptions), nodeResolve({ jsnext: true })].concat(
    compress ? uglify() : []
  ),
  moduleName: 'i18nextAsyncStorageBackend',
  // moduleId: 'i18nextAsyncStorageCache',
  globals: { '@react-native-async-storage/async-storage': 'AsyncStorage' },
  dest,
  external: ['@react-native-async-storage/async-storage']
};


i18nextAsyncStorageBackend.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('@react-native-async-storage/async-storage')) :
	typeof define === 'function' && define.amd ? define(['@react-native-async-storage/async-storage'], factory) :
	(global.i18nextAsyncStorageBackend = factory(global.AsyncStorage));
}(this, (function (AsyncStorage) { 'use strict';

AsyncStorage = 'default' in AsyncStorage ? AsyncStorage['default'] : AsyncStorage;

var arr = [];
var each = arr.forEach;
var slice = arr.slice;

function defaults(obj) {
  each.call(slice.call(arguments, 1), function (source) {
    if (source) {
      for (var prop in source) {
        if (obj[prop] === undefined) obj[prop] = source[prop];
      }
    }
  });
  return obj;
}

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// get from whatever version of react native that is being used.

var storage = {
  setItem: function setItem(key, value) {
    if (AsyncStorage) {
      return AsyncStorage.setItem(key, value);
    }
  },
  getItem: function getItem(key, value) {
    if (AsyncStorage) {
      return AsyncStorage.getItem(key, value);
    }
    return undefined;
  }
};

function getDefaults() {
  return {
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

var Cache = function () {
  function Cache(services) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Cache);

    this.init(services, options);

    this.type = 'backend';
  }

  _createClass(Cache, [{
    key: 'init',
    value: function init(services) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      this.services = services;
      this.options = defaults(options, this.options || {}, getDefaults());
    }
  }, {
    key: 'read',
    value: function read(language, namespace, callback) {
      var _this = this;

      var nowMS = new Date().getTime();

      if (!AsyncStorage) {
        return callback(null, null);
      }

      storage.getItem('' + this.options.prefix + language + '-' + namespace).then(function (local) {
        if (local) {
          local = JSON.parse(local);
          if (
          // expiration field is mandatory, and should not be expired
          local.i18nStamp && local.i18nStamp + _this.options.expirationTime > nowMS &&
          // there should be no language version set, or if it is, it should match the one in translation
          _this.options.versions[language] === local.i18nVersion) {
            delete local.i18nVersion;
            delete local.i18nStamp;
            return callback(null, local);
          }
        }

        callback(null, null);
      }).catch(function (err) {
        console.warn(err);
        callback(null, null);
      });
    }
  }, {
    key: 'save',
    value: function save(language, namespace, data) {
      if (AsyncStorage) {
        data.i18nStamp = new Date().getTime();

        // language version (if set)
        if (this.options.versions[language]) {
          data.i18nVersion = this.options.versions[language];
        }

        // save
        storage.setItem('' + this.options.prefix + language + '-' + namespace, JSON.stringify(data));
      }
    }
  }]);

  return Cache;
}();

Cache.type = 'backend';

return Cache;

})));
